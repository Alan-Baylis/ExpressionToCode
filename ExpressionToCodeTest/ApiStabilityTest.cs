using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using ExpressionToCodeLib;
using Xunit;

namespace ExpressionToCodeTest
{
    public class ApiStabilityTest
    {
        [Fact]
        public void PublicApi()
        {
            var publicTypes = typeof(ExpressionToCode).Assembly.GetTypes()
                .Where(IsPublic)
                .Where(type => !type.Namespace.Contains("Unstable"))
                .OrderByDescending(type => type.IsEnum)
                .ThenByDescending(type => type.IsInterface)
                .ThenBy(type => type.FullName);

            ApprovalTest.Verify(PrettyPrintTypes(publicTypes));
        }

        [Fact]
        public void UnstableApi()
        {
            var unstableTypes = typeof(ExpressionToCode).Assembly.GetTypes()
                .Where(IsPublic)
                .Where(type => type.Namespace.Contains("Unstable"))
                .OrderByDescending(type => type.IsEnum)
                .ThenByDescending(type => type.IsInterface)
                .ThenBy(type => type.FullName);

            ApprovalTest.Verify(PrettyPrintTypes(unstableTypes));
        }

        static string PrettyPrintTypes(IEnumerable<Type> types) => string.Join("", types.Select(PrettyPrintTypeDescription));
        static string PrettyPrintTypeDescription(Type o) => PrettyPrintTypeHeader(o) + "\n" + PrettyPrintTypeContents(o);

        static string PrettyPrintTypeContents(Type type)
        {
            var methods = type.GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
                    .OrderBy(mi => mi.MetadataToken)
                    .Where(mi => mi.DeclaringType.Assembly != typeof(object).Assembly) //exclude noise
                ;

            var methodBlock = string.Join("", methods.Select(mi => PrettyPrintMethod(mi) + "\n"));

            var fields = type.GetFields(BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
                    .Where(mi => mi.DeclaringType.Assembly != typeof(object).Assembly) //exclude noise
                ;

            var fieldBlock = string.Join("", fields.Select(fi => PrettyPrintField(fi) + "\n"));

            return fieldBlock + methodBlock + "\n";
        }

        static string PrettyPrintTypeHeader(Type type)
        {
            var prefix = TypePrefix(type);

            var baseType = type.BaseType == typeof(object) ? null : type.BaseType;
            var allInterfaces = type.GetInterfaces();
            var interfaces = baseType == null ? allInterfaces : allInterfaces.Except(baseType.GetInterfaces());
            var inheritanceTypes = new[] { baseType }.OfType<Type>().Concat(interfaces);
            var suffix = !inheritanceTypes.Any() || type.IsEnum ? "" : " : " + string.Join(", ", inheritanceTypes.Select(ObjectToCode.ToCSharpFriendlyTypeName));

            var name = ObjectToCode.ToCSharpFriendlyTypeName(type);

            return prefix + " " + name + suffix;
        }

        static string TypePrefix(Type type)
        {
            if (type.IsEnum) {
                return "enum";
            } else if (type.IsValueType) {
                return "struct";
            } else if (type.IsInterface) {
                return "interface";
            } else {
                return "class";
            }
        }

        static string PrettyPrintMethod(MethodInfo mi)
        {
            var fakeTarget = mi.IsStatic ? "TYPE" : "inst";

            return "    " + ObjectToCode.ToCSharpFriendlyTypeName(mi.ReturnType) + " " + fakeTarget +
                "." + mi.Name
                + PrettyPrintGenericArguments(mi)
                + PrettyPrintParameterList(mi);
        }

        static object PrettyPrintField(FieldInfo fi) => "    "
            + (fi.IsLiteral ? "const " : (fi.IsStatic ? "static " : "") + (fi.IsInitOnly ? "readonly " : ""))
            + ObjectToCode.ToCSharpFriendlyTypeName(fi.FieldType)
            + " " + fi.Name
            + (fi.IsLiteral ? " = " + ObjectToCode.ComplexObjectToPseudoCode(fi.GetRawConstantValue()) : "");

        static string PrettyPrintParameterList(MethodInfo mi)
        {
            return "(" + string.Join(
                ", ",
                mi.GetParameters()
                    .Select(
                        pi =>
                            ObjectToCode.ToCSharpFriendlyTypeName(pi.ParameterType) + " " + pi.Name)) + ")";
        }

        static string PrettyPrintGenericArguments(MethodInfo mi)
        {
            if (!mi.IsGenericMethodDefinition) {
                return "";
            }
            return "<"
                + string.Join(", ", mi.GetGenericArguments().Select(ObjectToCode.ToCSharpFriendlyTypeName))
                + ">";
        }

        static bool IsPublic(Type type) => type.IsPublic || type.IsNestedPublic && IsPublic(type.DeclaringType);
    }
}
