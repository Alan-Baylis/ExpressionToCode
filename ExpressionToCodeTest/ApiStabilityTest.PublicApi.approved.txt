enum EqualityExpressionClass
    int value__
    const EqualityExpressionClass None = 0
    const EqualityExpressionClass EqualsOp = 1
    const EqualityExpressionClass NotEqualsOp = 2
    const EqualityExpressionClass ObjectEquals = 3
    const EqualityExpressionClass ObjectEqualsStatic = 4
    const EqualityExpressionClass ObjectReferenceEquals = 5
    const EqualityExpressionClass EquatableEquals = 6
    const EqualityExpressionClass SequenceEqual = 7
    const EqualityExpressionClass StructuralEquals = 8

interface IExpressionTypeDispatch<T>
    T inst.DispatchAdd(Expression e)
    T inst.DispatchAddChecked(Expression e)
    T inst.DispatchAnd(Expression e)
    T inst.DispatchAndAlso(Expression e)
    T inst.DispatchArrayLength(Expression e)
    T inst.DispatchArrayIndex(Expression e)
    T inst.DispatchCall(Expression e)
    T inst.DispatchCoalesce(Expression e)
    T inst.DispatchConditional(Expression e)
    T inst.DispatchConstant(Expression e)
    T inst.DispatchConvert(Expression e)
    T inst.DispatchConvertChecked(Expression e)
    T inst.DispatchDivide(Expression e)
    T inst.DispatchEqual(Expression e)
    T inst.DispatchExclusiveOr(Expression e)
    T inst.DispatchGreaterThan(Expression e)
    T inst.DispatchGreaterThanOrEqual(Expression e)
    T inst.DispatchInvoke(Expression e)
    T inst.DispatchLambda(Expression e)
    T inst.DispatchLeftShift(Expression e)
    T inst.DispatchLessThan(Expression e)
    T inst.DispatchLessThanOrEqual(Expression e)
    T inst.DispatchListInit(Expression e)
    T inst.DispatchMemberAccess(Expression e)
    T inst.DispatchMemberInit(Expression e)
    T inst.DispatchModulo(Expression e)
    T inst.DispatchMultiply(Expression e)
    T inst.DispatchMultiplyChecked(Expression e)
    T inst.DispatchNegate(Expression e)
    T inst.DispatchUnaryPlus(Expression e)
    T inst.DispatchNegateChecked(Expression e)
    T inst.DispatchNew(Expression e)
    T inst.DispatchNewArrayInit(Expression e)
    T inst.DispatchNewArrayBounds(Expression e)
    T inst.DispatchNot(Expression e)
    T inst.DispatchNotEqual(Expression e)
    T inst.DispatchOr(Expression e)
    T inst.DispatchOrElse(Expression e)
    T inst.DispatchParameter(Expression e)
    T inst.DispatchPower(Expression e)
    T inst.DispatchQuote(Expression e)
    T inst.DispatchRightShift(Expression e)
    T inst.DispatchSubtract(Expression e)
    T inst.DispatchSubtractChecked(Expression e)
    T inst.DispatchTypeAs(Expression e)
    T inst.DispatchTypeIs(Expression e)
    T inst.DispatchAssign(Expression e)
    T inst.DispatchBlock(Expression e)
    T inst.DispatchDebugInfo(Expression e)
    T inst.DispatchDecrement(Expression e)
    T inst.DispatchDynamic(Expression e)
    T inst.DispatchDefault(Expression e)
    T inst.DispatchExtension(Expression e)
    T inst.DispatchGoto(Expression e)
    T inst.DispatchIncrement(Expression e)
    T inst.DispatchIndex(Expression e)
    T inst.DispatchLabel(Expression e)
    T inst.DispatchRuntimeVariables(Expression e)
    T inst.DispatchLoop(Expression e)
    T inst.DispatchSwitch(Expression e)
    T inst.DispatchThrow(Expression e)
    T inst.DispatchTry(Expression e)
    T inst.DispatchUnbox(Expression e)
    T inst.DispatchAddAssign(Expression e)
    T inst.DispatchAndAssign(Expression e)
    T inst.DispatchDivideAssign(Expression e)
    T inst.DispatchExclusiveOrAssign(Expression e)
    T inst.DispatchLeftShiftAssign(Expression e)
    T inst.DispatchModuloAssign(Expression e)
    T inst.DispatchMultiplyAssign(Expression e)
    T inst.DispatchOrAssign(Expression e)
    T inst.DispatchPowerAssign(Expression e)
    T inst.DispatchRightShiftAssign(Expression e)
    T inst.DispatchSubtractAssign(Expression e)
    T inst.DispatchAddAssignChecked(Expression e)
    T inst.DispatchMultiplyAssignChecked(Expression e)
    T inst.DispatchSubtractAssignChecked(Expression e)
    T inst.DispatchPreIncrementAssign(Expression e)
    T inst.DispatchPreDecrementAssign(Expression e)
    T inst.DispatchPostIncrementAssign(Expression e)
    T inst.DispatchPostDecrementAssign(Expression e)
    T inst.DispatchTypeEqual(Expression e)
    T inst.DispatchOnesComplement(Expression e)
    T inst.DispatchIsTrue(Expression e)
    T inst.DispatchIsFalse(Expression e)

class EqualityExpressions
    EqualityExpressionClass TYPE.CheckForEquality(Expression<Func<bool>> e)
    Tuple<EqualityExpressionClass, Expression, Expression> TYPE.ExtractEqualityType(Expression<Func<bool>> e)
    Tuple<EqualityExpressionClass, Expression, Expression> TYPE.ExtractEqualityType(Expression e)
    IEnumerable<Tuple<EqualityExpressionClass, bool>> TYPE.DisagreeingEqualities(Expression<Func<bool>> e)
    IEnumerable<Tuple<EqualityExpressionClass, bool>> TYPE.DisagreeingEqualities(Expression left, Expression right, bool shouldBeEqual)

class ExpressionToCode
    string TYPE.ToCode<T, T1, T2, T3>(Expression<Func<T, T1, T2, T3>> e)
    string TYPE.ToCode<T, T1, T2>(Expression<Func<T, T1, T2>> e)
    string TYPE.ToCode<T, T1>(Expression<Func<T, T1>> e)
    string TYPE.ToCode<T>(Expression<Func<T>> e)
    string TYPE.AnnotatedToCode<T, T1, T2, T3>(Expression<Func<T, T1, T2, T3>> e)
    string TYPE.AnnotatedToCode<T, T1, T2>(Expression<Func<T, T1, T2>> e)
    string TYPE.AnnotatedToCode<T, T1>(Expression<Func<T, T1>> e)
    string TYPE.AnnotatedToCode<T>(Expression<Func<T>> e)
    string TYPE.ToCode(Expression e)
    string TYPE.AnnotatedToCode(Expression expr)

class ExpressionTypeDispatcher
    T TYPE.ExpressionDispatch<T>(IExpressionTypeDispatch<T> dispatcher, Expression e)

class UnitTestingInternalsAccess
    Exception TYPE.CreateException(string msg)

class ObjectToCode
    string TYPE.ComplexObjectToPseudoCode(object val, int indent)
    string TYPE.PlainObjectToCode(object val)
    string TYPE.PlainObjectToCode(object val, Type type)
    string TYPE.GetCSharpFriendlyTypeName(Type type)

class PAssert
    void TYPE.IsTrue(Expression<Func<bool>> assertion)
    void TYPE.That(Expression<Func<bool>> assertion, string msg)

class PAssertFailedException : Exception

